firedrake/mg/mesh.py

```diff
@@ -102,6 +108,13 @@ def MeshHierarchy(mesh, refinement_levels,
         callback receives the refined DM (and the current level).
     :arg mesh_builder: Function to turn a DM into a ``Mesh``. Used by pyadjoint.
     """
+    if netgen and isinstance(mesh, netgen.libngpy._meshing.Mesh):
+        try:
+            from ngsPETSc import NetgenHierarchy
+        except ImportError:
+            raise ImportError("Unable to import ngsPETSc. Please ensure that ngsolve is installed and available to Firedrake.")
+        return NetgenHierarchy(mesh, refinement_levels, order=order, digits=8, adaptive=False)
```

Comment:
I don't really understand what you have done. This completely avoids a lot of important Firedrake mesh logic. How does this work?

Reply:
Everything is hidden inside of ngsPETSc.
More detail here: https://github.com/NGSolve/ngsPETSc/blob/main/ngsPETSc/utils/firedrake.py